### PR TITLE
Task04 Даниил Ушков ITMO

### PR DESCRIPTION
**Pull request by daniil-ushkov**
<details><summary>Локальный вывод</summary><p>

<pre>
OpenCL devices:
  Device #0: CPU. 12th Gen Intel(R) Core(TM) i7-1260P. Intel(R) Corporation. Total memory: 31716 Mb
Using device #0: CPU. 12th Gen Intel(R) Core(TM) i7-1260P. Intel(R) Corporation. Total memory: 31716 Mb
Data generated for M=4096, K=4096
[matrix_transpose_naive]
    GPU: 0.00643078+-7.0743e-05 s
    GPU: 2608.89 millions/s
[matrix_transpose_local_bad_banks]
    GPU: 0.00673028+-7.79243e-05 s
    GPU: 2492.79 millions/s
[matrix_transpose_local_good_banks]
    GPU: 0.00682775+-0.000100604 s
    GPU: 2457.21 millions/s

OpenCL devices:
  Device #0: CPU. 12th Gen Intel(R) Core(TM) i7-1260P. Intel(R) Corporation. Total memory: 31716 Mb
Using device #0: CPU. 12th Gen Intel(R) Core(TM) i7-1260P. Intel(R) Corporation. Total memory: 31716 Mb
Data generated for M=1024, K=1024, N=1024
CPU: 2.86558+-0 s
CPU: 0.69794 GFlops
[naive, ts=4]
    GPU: 0.132822+-0.00214371 s
    GPU: 15.0577 GFlops
    Average difference: 0.000149043%
[naive, ts=8]
    GPU: 0.122077+-0.00402763 s
    GPU: 16.383 GFlops
    Average difference: 0.000149043%
[naive, ts=16]
    GPU: 0.122976+-0.00296021 s
    GPU: 16.2634 GFlops
    Average difference: 0.000149043%
[local, ts=4]
    GPU: 0.194005+-0.00112202 s
    GPU: 10.309 GFlops
    Average difference: 0.000149043%
[local, ts=8]
    GPU: 0.11476+-0.000601609 s
    GPU: 17.4276 GFlops
    Average difference: 0.000149043%
[local, ts=16]
    GPU: 0.0940113+-0.000430166 s
    GPU: 21.274 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=2]
    GPU: 0.159422+-0.00145472 s
    GPU: 12.5453 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.129037+-0.000798055 s
    GPU: 15.4994 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.101135+-0.00340062 s
    GPU: 19.7755 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=4]
    GPU: 0.0864793+-0.000784055 s
    GPU: 23.1269 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=8]
    GPU: 0.075373+-0.00124269 s
    GPU: 26.5347 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=2]
    GPU: 0.0714007+-0.000784316 s
    GPU: 28.0109 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=4]
    GPU: 0.0584528+-0.000802348 s
    GPU: 34.2156 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=8]
    GPU: 0.054619+-0.000709091 s
    GPU: 36.6173 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=16]
    GPU: 0.0768372+-0.000169198 s
    GPU: 26.0291 GFlops
    Average difference: 0.000149043%
</p></details>

<details><summary>Вывод Github CI</summary><p>

<pre>
OpenCL devices:
  Device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Using device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Data generated for M=4096, K=4096
[matrix_transpose_naive]
    GPU: 0.00955785+-5.10574e-05 s
    GPU: 1755.33 millions/s
[matrix_transpose_local_bad_banks]
    GPU: 0.0160714+-9.57283e-05 s
    GPU: 1043.92 millions/s
[matrix_transpose_local_good_banks]
    GPU: 0.0159575+-9.54326e-05 s
    GPU: 1051.37 millions/s

OpenCL devices:
  Device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Using device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Data generated for M=1024, K=1024, N=1024
CPU: 6.40108+-0 s
CPU: 0.312447 GFlops
[naive, ts=4]
    GPU: 0.308028+-0.00499184 s
    GPU: 6.49292 GFlops
    Average difference: 0.000149043%
[naive, ts=8]
    GPU: 0.363087+-0.00546321 s
    GPU: 5.50832 GFlops
    Average difference: 0.000149043%
[naive, ts=16]
    GPU: 0.353934+-0.00252341 s
    GPU: 5.65077 GFlops
    Average difference: 0.000149043%
[local, ts=4]
    GPU: 0.614314+-0.00221631 s
    GPU: 3.25566 GFlops
    Average difference: 0.000149043%
[local, ts=8]
    GPU: 0.400141+-0.000591302 s
    GPU: 4.99824 GFlops
    Average difference: 0.000149043%
[local, ts=16]
    GPU: 0.301455+-0.000271772 s
    GPU: 6.63449 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=2]
    GPU: 0.532485+-0.000985548 s
    GPU: 3.75598 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.475623+-0.00114181 s
    GPU: 4.20501 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.35216+-0.000416518 s
    GPU: 5.67924 GFlops
    Average difference: 0.000149043%
</pre>

</p></details>

### CODE REVIEW

**Comment by daniil-ushkov** [comment]
Анализ производительности в `ANSWER.md`

File: src/cl/matrix_multiplication.cl [hunk 1]
```diff
@@ -7,21 +7,123 @@
 
 // TILE_SIZE и WORK_PER_THREAD задаются через поле 'defines' в кернел конфиге
 
-__kernel void matrix_multiplication_naive()
+__kernel void matrix_multiplication_naive(__global float *a,
```
Comment (simiyutin) [inline src/cl/matrix_multiplication.cl:10]: не коалесд чтение и запись из глобальной памяти

File: src/cl/matrix_multiplication.cl [hunk 2]
```diff
@@ -7,21 +7,123 @@
...
+    for (unsigned int k = 0; k < K; k++) {
+        sum += a[K * i + k] * b[N * k + j];
+    }
+
+    c[N * i + j] = sum;
 }
 
 #ifdef TILE_SIZE
-__kernel void matrix_multiplication_local()
+__kernel void matrix_multiplication_local(__global float *a,
```
Comment (simiyutin) [inline src/cl/matrix_multiplication.cl:33]: не коалесд чтение и запись из глобальной памяти

File: src/cl/matrix_multiplication.cl [hunk 3]
```diff
@@ -7,21 +7,123 @@
...
+
+    if (i < M && j < N) {
+        c[N * i + j] = sum;
+    }
 }
 #endif
 
 #if defined(TILE_SIZE) && defined(WORK_PER_THREAD)
-__kernel void matrix_multiplication_local_wpt()
+__kernel void matrix_multiplication_local_wpt(__global float *a,
```
Comment (simiyutin) [inline src/cl/matrix_multiplication.cl:76]: не коалесд чтение и запись из глобальной памяти

File: src/cl/matrix_transpose.cl [hunk 4]
```diff
@@ -5,17 +5,45 @@
...
-    // TODO
+    unsigned int i = get_global_id(0);
+    unsigned int j = get_global_id(1);
+
+    __local float tile[TILE_SIZE][TILE_SIZE];
+
+    unsigned int local_i = get_local_id(0);
+    unsigned int local_j = get_local_id(1);
+
+    tile[local_i][local_j] = as[K * i + j];
```
Comment (simiyutin) [inline src/cl/matrix_transpose.cl:27]: Не коалесд доступ к глобальной памяти + в таком использовании локального массива нет смысла, поток использует то же значение что сам и загрузил

File: src/cl/matrix_transpose.cl [hunk 5]
```diff
@@ -5,17 +5,45 @@
...
-    // TODO
+    unsigned int i = get_global_id(0);
+    unsigned int j = get_global_id(1);
+
+    __local float tile[TILE_SIZE + 1][TILE_SIZE];
+
+    unsigned int local_i = get_local_id(0);
+    unsigned int local_j = get_local_id(1);
+
+    tile[local_i][local_j] = as[K * i + j];
```
Comment (simiyutin) [inline src/cl/matrix_transpose.cl:44]: и здесь

File: src/main_matrix_multiplication.cpp [hunk 6]
```diff
@@ -69,7 +69,7 @@ KernelConfig makeLocalConfig(unsigned int tile_size)
 KernelConfig makeLocalWPTConfig(unsigned int tile_size, unsigned int wpt)
 {
     std::string kernel_name = "matrix_multiplication_local_wpt";
-    gpu::WorkSize work_size(tile_size, tile_size / wpt, M, N / wpt);
+    gpu::WorkSize work_size(tile_size / wpt, tile_size, N / wpt, M);
```
Comment (simiyutin) [inline src/main_matrix_multiplication.cpp:72]: теперь здесь появился скрытый не-коалесд доступ, когда сжимаем рабочее пространство и рабочую группу по строчкам, то один варп грузит только по `tile_size/wpt` элементов на каждой итерации цикла по `w`

Comment (simiyutin) [reply src/main_matrix_multiplication.cpp:72]: из слайдов к лекции:

image

Comment (daniil-ushkov) [reply src/main_matrix_multiplication.cpp]: Поправил, но производительность на cpu упала, а gpu чтобы потестировать сейчас нет рядом =(
```
[local wpt, ts=4, wpt=2]
    GPU: 0.225491+-0.00513369 s
    GPU: 8.86953 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.177312+-0.00336677 s
    GPU: 11.2796 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.105111+-0.00275048 s
    GPU: 19.0275 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=4]
    GPU: 0.0970787+-0.00357226 s
    GPU: 20.6018 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=8]
    GPU: 0.0924922+-0.00503864 s
    GPU: 21.6235 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=2]
    GPU: 0.0771242+-0.00215991 s
    GPU: 25.9322 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=4]
    GPU: 0.0665693+-0.0015366 s
    GPU: 30.0439 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=8]
    GPU: 0.0697818+-0.00189218 s
    GPU: 28.6608 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=16]
    GPU: 0.104678+-0.00469397 s
    GPU: 19.1062 GFlops
    Average difference: 0.000149043%
```

Comment (daniil-ushkov) [reply src/main_matrix_multiplication.cpp]: Посмотрите еще раз?

Comment (simiyutin) [reply src/main_matrix_multiplication.cpp]: Да, теперь вроде бы все хорошо)

**Comment by simiyutin** [comment]
Все хорошо, задача зачтена, 10/10 баллов :+1:

**Pull request closed without merging**